### PR TITLE
Google's version of virtio drivers.

### PR DESCRIPTION
**Pull request by iuliur**
Google's version of virtio for reference.

### CODE REVIEW

**Comment by ladipro** [comment]
This is huge! :) Give me some time to go over the changes. Thanks!

File: vioscsi/vioscsi.c [hunk 1]
```diff
@@ -417,16 +420,22 @@ ENTER_FN();
...
                 SP_INTERNAL_ADAPTER_ERROR,
                 __LINE__);
-
-            RhelDbgPrint(TRACE_LEVEL_FATAL, ("Virtual queue %d config failed.\n", index));
+            TRACE1(TRACE_LEVEL_FATAL, DRIVER_START, "Virtual queue config failed", "queue", index);
             return SP_RETURN_ERROR;
         }
+        if (adaptExt->dump_mode && queueLength > MAX_DUMP_MODE_QUEUE_NUM) {
+           adaptExt->original_queue_num[index] = queueLength;
+           virtio_set_queue_allocation(&adaptExt->vdev, index, MAX_DUMP_MODE_QUEUE_NUM);
```
Comment (ladipro) [inline vioscsi/vioscsi.c:435]: I finally got around to looking at this in detail.

So I see that you limit the queue size to 256 when running in dump mode. What issue prompted this if I may ask? Problems allocating virtqueue memory? Why not initialize only one queue instead?

If I'm reading the code right, this:

`      max_queues = min(max_cpus, adaptExt->scsi_config.num_queues);`

means that max_queues is potentially a large number, even in dump mode.

**Comment by iuliur** [comment]
On 2008, storportgetunchachedextension fails to allocate more than 64k in
dump mode. Try it on your side - have only one queue but make it like 4k
entries. If you crash the machine the dump file won't be saved as adapter
initialization fails.

**Comment by ladipro** [comment]
Got it, yes. Our hypervisor has vioscsi queue sizes fixed to 128 entries so the code you've added is not needed for us. But I can reproduce this on Win7 even with num_queues=4 and our 128-entry queues so I am going to post a fix. Thanks!

**Comment by iuliur** [comment]
But dump mode always sets num queues to 1. It should not matter if you have
4 queues or not. Only the number of entries should matter. Unless something
else regressed about dump mode

**Comment by ladipro** [comment]
Take a look at where max_queues comes from. It is currently 4 in a 4-queue VM even in dump mode. I believe that your code has the bug too.

**Comment by ladipro** [comment]
https://github.com/virtio-win/kvm-guest-drivers-windows/pull/110

**Pull request closed without merging**